### PR TITLE
Add ID scraper

### LINKED ISSUE #82: Add ID scrape
https://www.labor.idaho.gov/dnn/Businesses/Layoff-Assistance#2

Links to a PDF that you can download.

### PR DESCRIPTION
**Pull request by chriszs**
Add Idaho scraper.

## Steps to test
```sh
python -m warn.cli ID
```

Closes #82

### CODE REVIEW

**Comment by chriszs** [comment]
Deals with gaps caused by merged cells, like this:

Screen Shot 2022-01-20 at 6 58 52 AM

By filling down like this:

Screen Shot 2022-01-21 at 7 57 20 PM

Notice we're excluding the number of employees column to preserve the merged cell total and avoid double counting.

File: warn/scrapers/id.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            clean_text = _clean_text(column)
+
+            # If cell is empty, copy from the cell above it
+            # to deal with merged cells. Except for number of employees,
+            # which is effectively a total for all locations in the merged cell
+            # and which we don't want a data user to double count.
+            if (
+                clean_text == ""
+                and row_index > 0
+                and col_index < len(output_rows[row_index - 1])
```
Comment (chriszs) [inline warn/scrapers/id.py:94]: Black and flake8 disagree about whether the `and` should go before the line or after, [this website says flake8 is wrong](https://www.flake8rules.com/rules/W503.html).

File: warn/scrapers/id.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    url = f"{base_url}{file_name}?v={cache_buster}"
+
+    cache_state = Path(cache_dir, state_code)
+    cache_state.mkdir(parents=True, exist_ok=True)
+
+    cache_key = f"{cache_state}/WARNNotice.pdf"
+
+    # verify=False because there's a persistent cert error
+    # we're working around.
```
Comment (chriszs) [inline warn/scrapers/id.py:50]: This results in a nasty warning, but as far as I can tell there may be no good way around it. I looked into whether the problem was not having `certifi` as [StackOverflow indicated](https://stackoverflow.com/questions/52805115/certificate-verify-failed-unable-to-get-local-issuer-certificate), but it seemed like that wasn't it.

File: warn/scrapers/id.py (hunk 1 above)
Comment (chriszs) [reply warn/scrapers/id.py:93]: Okay, upon researching this I realized that the linter was also complaining when I put these at the end of the line. How was it possible it didn't like _both_? And why does the flake8 maintainer say this has rule is ignored? Answer, using `ignore` in the flake8 config overrides default ignore rules, including this one. [This StackOverflow answer](https://stackoverflow.com/a/68172142/1871478) recommends `extend-ignore` instead, which is what this patch now modifies the flake8 config to do.

File: warn/scrapers/id.py (hunk 2 above)
Comment (chriszs) [reply warn/scrapers/id.py:49]: Leaving this as-is.

**Comment by chriszs** [comment]
This one's ready for review, @palewire.

File: warn/scrapers/id.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    url = f"{base_url}{file_name}?v={cache_buster}"
+
+    cache_state = Path(cache_dir, state_code)
+    cache_state.mkdir(parents=True, exist_ok=True)
+
+    cache_key = f"{cache_state}/{file_name}"
+
+    # verify=False because there's a persistent cert error
+    # we're working around.
+    response = requests.get(url, verify=False)
```
Comment (palewire) [inline warn/scrapers/id.py:50]: What do you think about adding a verify keyword option to our `utils.get_url` utility so we can use it here? We'd get the logging, and make the utility stronger too.

File: warn/scrapers/id.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    with open(cache_key, "wb") as file:
+        file.write(response.content)
```
Comment (palewire) [inline warn/scrapers/id.py:52]: Could we use the built-in `cache.write` function from the instance of our Cache class?

File: warn/scrapers/id.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+    output_rows: list = []
+
+    with pdfplumber.open(cache_key) as pdf:
+        for index, page in enumerate(pdf.pages):
+            rows = page.extract_table()
+
+            output_rows = output_rows + _clean_table(rows, index)
```
Comment (palewire) [inline warn/scrapers/id.py:60]:
```suggestion
            output_rows +=  _clean_table(rows, index)
```
I've actually never written a line like this, but would it work?!

File: warn/scrapers/id.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+    # Write out the data to a CSV
+    data_path = Path(data_dir, f"{state_code}.csv")
+    utils.write_rows_to_csv(output_rows, data_path)
```
Comment (palewire) [inline warn/scrapers/id.py:64]:
```suggestion
    utils.write_rows_to_csv(data_path, output_rows)
```
This got reversed upstream. Need to catch up.

File: warn/scrapers/id.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    output_rows: list = []
+
+    with pdfplumber.open(cache_key) as pdf:
+        for index, page in enumerate(pdf.pages):
+            rows = page.extract_table()
+
+            output_rows = output_rows + _clean_table(rows, index)
+
+    # Write out the data to a CSV
+    data_path = Path(data_dir, f"{state_code}.csv")
```
Comment (palewire) [inline warn/scrapers/id.py:63]:
```suggestion
    data_path = data_dir / f"{state_code}.csv"
```
My picky thing.

File: warn/scrapers/id.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+            # If cell is empty, copy from the cell above it
+            # to deal with merged cells. Except for number of employees,
+            # which is effectively a total for all locations in the merged cell
+            # and which we don't want a data user to double count.
+            if (
+                clean_text == ""
+                and row_index > 0
+                and col_index < len(output_rows[row_index - 1])
+                and output_rows[0][col_index] != "No. of Employees Affected"
```
Comment (palewire) [inline warn/scrapers/id.py:94]: Would you mind breaking this out into an `_is_header` true/false test so it's a little clear to newbies what the test is doing. (Assume I am clear what it's doing!)

**Review by palewire — CHANGES_REQUESTED** [review]
Thanks for the creative use pdfplumber here. Big picture it all looks great. I just have a couple small things I've flagged. Let me know what you think.

File: warn/scrapers/id.py (hunk 3 above)
Comment (chriszs) [reply warn/scrapers/id.py:50]: Yes, I think that's a good idea. I'll do it.

File: warn/scrapers/id.py (hunk 4 above)
Comment (chriszs) [reply warn/scrapers/id.py:52]: I tried to, but I think `cache.write` doesn't handle binary files. It only works for strings. The better approach here I guess would be to extend `cache.download` to take `verify` and use that.

File: warn/scrapers/id.py (hunk 8 above)
Comment (chriszs) [reply warn/scrapers/id.py]: It doesn't do that, but refactored into functions that say what it is doing.

File: warn/scrapers/id.py (hunk 4 above)
Comment (chriszs) [reply warn/scrapers/id.py]: Added.

**Comment by chriszs** [comment]
Thanks for the suggestions! Refactored to use `cache.download` (and added a feature to that to pass verify or any other kwarg), added some util functions to make that one if statement test read clearer, integrated your suggestions and cleaned everything up.

**Comment by palewire** [comment]
Thanks! You're in!

**Pull request merged**